File: core/src/main/java/com/netflix/iceberg/TableMetadataParser.java
/*
 * Copyright 2017 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.netflix.iceberg;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.netflix.iceberg.TableMetadata.SnapshotLogEntry;
import com.netflix.iceberg.exceptions.RuntimeIOException;
import com.netflix.iceberg.io.InputFile;
import com.netflix.iceberg.io.OutputFile;
import com.netflix.iceberg.util.JsonUtil;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.StringWriter;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.NavigableSet;
import java.util.SortedSet;
import java.util.TreeSet;

public class TableMetadataParser {

  private static final String FORMAT_VERSION = "format-version";
  private static final String LOCATION = "location";
  private static final String LAST_UPDATED_MILLIS = "last-updated-ms";
  private static final String LAST_COLUMN_ID = "last-column-id";
  private static final String SCHEMA = "schema";
  private static final String PARTITION_SPEC = "partition-spec";
  private static final String PROPERTIES = "properties";
  private static final String CURRENT_SNAPSHOT_ID = "current-snapshot-id";
  private static final String SNAPSHOTS = "snapshots";
  private static final String SNAPSHOT_ID = "snapshot-id";
  private static final String TIMESTAMP_MS = "timestamp-ms";
  private static final String SNAPSHOT_LOG = "snapshot-log";

  public static String toJson(TableMetadata metadata) {
    StringWriter writer = new StringWriter();
    try {
      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
      toJson(metadata, generator);
      generator.flush();
    } catch (IOException e) {
      throw new RuntimeIOException(e, "Failed to write json for: %s", metadata);
    }
    return writer.toString();
  }

  public static void write(TableMetadata metadata, OutputFile outputFile) {
    try (OutputStreamWriter writer = new OutputStreamWriter(outputFile.create())) {
      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
      generator.useDefaultPrettyPrinter();
      toJson(metadata, generator);
      generator.flush();
    } catch (IOException e) {
      throw new RuntimeIOException(e, "Failed to write json to file: %s", outputFile);
    }
  }

  private static void toJson(TableMetadata metadata, JsonGenerator generator) throws IOException {
    generator.writeStartObject();

    generator.writeNumberField(FORMAT_VERSION, TableMetadata.TABLE_FORMAT_VERSION);
    generator.writeStringField(LOCATION, metadata.location());
    generator.writeNumberField(LAST_UPDATED_MILLIS, metadata.lastUpdatedMillis());
    generator.writeNumberField(LAST_COLUMN_ID, metadata.lastColumnId());

    generator.writeFieldName(SCHEMA);
    SchemaParser.toJson(metadata.schema(), generator);

    generator.writeFieldName(PARTITION_SPEC);
    PartitionSpecParser.toJson(metadata.spec(), generator);

    generator.writeObjectFieldStart(PROPERTIES);
    for (Map.Entry<String, String> keyValue : metadata.properties().entrySet()) {
      generator.writeStringField(keyValue.getKey(), keyValue.getValue());
    }
    generator.writeEndObject();

    generator.writeNumberField(CURRENT_SNAPSHOT_ID,
        metadata.currentSnapshot() != null ? metadata.currentSnapshot().snapshotId() : -1);

    generator.writeArrayFieldStart(SNAPSHOTS);
    for (Snapshot snapshot : metadata.snapshots()) {
      SnapshotParser.toJson(snapshot, generator);
    }
    generator.writeEndArray();

    generator.writeArrayFieldStart(SNAPSHOT_LOG);
    for (SnapshotLogEntry logEntry : metadata.snapshotLog()) {
      generator.writeStartObject();
      generator.writeNumberField(TIMESTAMP_MS, logEntry.timestampMillis());
      generator.writeNumberField(SNAPSHOT_ID, logEntry.snapshotId());
      generator.writeEndObject();
    }
    generator.writeEndArray();

    generator.writeEndObject();
  }

  public static TableMetadata read(TableOperations ops, InputFile file) {
    try {
      return fromJson(ops, file, JsonUtil.mapper().readValue(file.newStream(), JsonNode.class));
    } catch (IOException e) {
      throw new RuntimeIOException(e, "Failed to read file: %s", file);
    }
  }

  static TableMetadata fromJson(TableOperations ops, InputFile file, JsonNode node) {
    Preconditions.checkArgument(node.isObject(),
        "Cannot parse metadata from a non-object: %s", node);

    int formatVersion = JsonUtil.getInt(FORMAT_VERSION, node);
    Preconditions.checkArgument(formatVersion == TableMetadata.TABLE_FORMAT_VERSION,
        "Cannot read unsupported version %d", formatVersion);

    String location = JsonUtil.getString(LOCATION, node);
    int lastAssignedColumnId = JsonUtil.getInt(LAST_COLUMN_ID, node);
    Schema schema = SchemaParser.fromJson(node.get(SCHEMA));
    PartitionSpec spec = PartitionSpecParser.fromJson(schema, node.get(PARTITION_SPEC));
    Map<String, String> properties = JsonUtil.getStringMap(PROPERTIES, node);
    long currentVersionId = JsonUtil.getLong(CURRENT_SNAPSHOT_ID, node);
    long lastUpdatedMillis = JsonUtil.getLong(LAST_UPDATED_MILLIS, node);

    JsonNode snapshotArray = node.get(SNAPSHOTS);
    Preconditions.checkArgument(snapshotArray.isArray(),
        "Cannot parse snapshots from non-array: %s", snapshotArray);

    List<Snapshot> snapshots = Lists.newArrayListWithExpectedSize(snapshotArray.size());
    Iterator<JsonNode> iterator = snapshotArray.elements();
    while (iterator.hasNext()) {
      snapshots.add(SnapshotParser.fromJson(ops, iterator.next()));
    }

    SortedSet<SnapshotLogEntry> entries =
        Sets.newTreeSet(Comparator.comparingLong(SnapshotLogEntry::timestampMillis));
    if (node.has(SNAPSHOT_LOG)) {
      Iterator<JsonNode> logIterator = node.get(SNAPSHOT_LOG).elements();
      while (logIterator.hasNext()) {
        JsonNode entryNode = logIterator.next();
        entries.add(new SnapshotLogEntry(
            JsonUtil.getLong(TIMESTAMP_MS, entryNode), JsonUtil.getLong(SNAPSHOT_ID, entryNode)));
      }
    }

    return new TableMetadata(ops, file, location,
        lastUpdatedMillis, lastAssignedColumnId, schema, spec, properties, currentVersionId,
        snapshots, ImmutableList.copyOf(entries.iterator()));
  }

}

File: python/iceberg/core/table_metadata_parser.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import gzip
import json

from .config_properties import ConfigProperties
from .partition_spec_parser import PartitionSpecParser
from .schema_parser import SchemaParser
from .snapshot_parser import SnapshotParser
from .table_metadata import (SnapshotLogEntry,
                             TableMetadata)


class TableMetadataParser(object):
    FORMAT_VERSION = "format-version"
    LOCATION = "location"
    LAST_UPDATED_MILLIS = "last-updated-ms"
    LAST_COLUMN_ID = "last-column-id"
    SCHEMA = "schema"
    PARTITION_SPEC = "partition-spec"
    PARTITION_SPECS = "partition-specs"
    DEFAULT_SPEC_ID = "default-spec-id"
    PROPERTIES = "properties"
    CURRENT_SNAPSHOT_ID = "current-snapshot-id"
    SNAPSHOTS = "snapshots"
    SNAPSHOT_ID = "snapshot-id"
    TIMESTAMP_MS = "timestamp-ms"
    LOG = "snapshot-log"

    @staticmethod
    def to_json(metadata, indent=4):
        return json.dumps({TableMetadataParser.FORMAT_VERSION: TableMetadata.TABLE_FORMAT_VERSION,
                           TableMetadataParser.LOCATION: metadata.location,
                           TableMetadataParser.LAST_UPDATED_MILLIS: metadata.last_updated_millis,
                           TableMetadataParser.LAST_COLUMN_ID: metadata.last_column_id,
                           TableMetadataParser.SCHEMA: SchemaParser.to_dict(metadata.schema),
                           TableMetadataParser.PARTITION_SPEC: PartitionSpecParser.to_json_fields(metadata.spec),
                           TableMetadataParser.DEFAULT_SPEC_ID: int(metadata.default_spec_id),
                           TableMetadataParser.PARTITION_SPECS: [PartitionSpecParser.to_dict(spec)
                                                                 for spec in metadata.specs],
                           TableMetadataParser.PROPERTIES: metadata.properties,
                           TableMetadataParser.CURRENT_SNAPSHOT_ID: (metadata.current_snapshot_id
                                                                     if metadata.current_snapshot_id is not None
                                                                     else -1),
                           TableMetadataParser.SNAPSHOTS: [SnapshotParser.to_dict(snapshot)
                                                           for snapshot in metadata.snapshots],
                           TableMetadataParser.LOG: [{TableMetadataParser.TIMESTAMP_MS: log_entry.timestamp_millis,
                                                      TableMetadataParser.SNAPSHOT_ID: log_entry.snapshot_id}
                                                     for log_entry in metadata.snapshot_log]}, indent=indent)

    @staticmethod
    def write(metadata, metadata_location):
        if metadata_location.location().endswith(".gz"):
            output_file = gzip.open(metadata_location.create("wb"), "wb")
        else:
            output_file = metadata_location.create("wb")

        json_str = TableMetadataParser.to_json(metadata)
        output_file.write(json_str.encode("utf-8"))
        output_file.close()

    @staticmethod
    def get_file_extension(config):
        return ".metadata.json.gz" if ConfigProperties.should_compress(config) else ".metadata.json"

    @staticmethod
    def read(ops, file):
        metadata = "".join([line.decode("utf-8") for line in file.new_stream(gzipped=file.location().endswith("gz"))])
        return TableMetadataParser.from_json(ops, file.location(), metadata)

    @staticmethod
    def from_json(ops, file, json_obj):
        if isinstance(json_obj, str):
            json_obj = json.loads(json_obj)

        if not isinstance(json_obj, dict):
            raise RuntimeError("Cannot parse metadata from non-object: %s" % json_obj)

        format_version = json_obj.get(TableMetadataParser.FORMAT_VERSION)
        if format_version != TableMetadata.TABLE_FORMAT_VERSION:
            raise RuntimeError("Cannot read unsupported version: %s" % format_version)

        location = json_obj.get(TableMetadataParser.LOCATION)
        last_assigned_column = json_obj.get(TableMetadataParser.LAST_COLUMN_ID)
        schema = SchemaParser.from_json(json_obj.get(TableMetadataParser.SCHEMA))

        spec_array = json_obj.get(TableMetadataParser.PARTITION_SPECS)
        if spec_array is not None:
            default_spec_id = json_obj.get(TableMetadataParser.DEFAULT_SPEC_ID)
            specs = [PartitionSpecParser.from_json(schema, spec)
                     for spec in spec_array]
        else:
            default_spec_id = TableMetadata.INITIAL_SPEC_ID
            specs = (PartitionSpecParser.from_json_fields(schema,
                                                          default_spec_id,
                                                          json_obj.get(TableMetadataParser.PARTITION_SPEC)),)

        props = json_obj.get(TableMetadataParser.PROPERTIES)
        current_version_id = json_obj.get(TableMetadataParser.CURRENT_SNAPSHOT_ID)
        last_updated_millis = json_obj.get(TableMetadataParser.LAST_UPDATED_MILLIS)
        snapshots = [SnapshotParser.from_json(ops, snapshot) for snapshot in json_obj.get(TableMetadataParser.SNAPSHOTS)]
        entries = [SnapshotLogEntry(log_entry.get(TableMetadataParser.TIMESTAMP_MS),
                                    log_entry.get(TableMetadataParser.SNAPSHOT_ID))
                   for log_entry in sorted(json_obj.get(TableMetadataParser.LOG, []),
                                           key=lambda x: x.get(TableMetadataParser.TIMESTAMP_MS))]

        return TableMetadata(ops, file, location,
                             last_updated_millis, last_assigned_column, schema, default_spec_id, specs, props, current_version_id,
                             snapshots, entries)
